### PR TITLE
Pp reshard develop

### PR DESCRIPTION
**Pull request by liuzhenhai93**
### PR types

Others
### PR changes

Others

##### 背景
大模型多卡分布式的训练保存的Checkpoint通常是根据张量并行、FSDP、流水线并行进行切分保存的。
在恢复模型训练时，可能会面临的分布式策略改变、训练节点数发生变化的情况。比如探索不同的混合并行策略，或切换训练任务（pretrain -> sft、rlhf）。在这种情况下，用户往往需要离线手动对Checkpoint进行处理，比较繁琐。
为了解决上述的问题，解放生产力，我们开发了模型热启自适应reshard功能。

##### 功能

模型热启自适应 reshard 功能支持：
    在同一个模型在其他配置不变的情况下, 任意切换4D分布式并行策略（MP、PP、SHARDING、DP）后，依然可以自适应分布式并行策略，恢复原来的模型状态量。这里的模型状态量不仅包括权重本身，还包括learning rate scheduler 的状态量以及优化器的状态量。

##### 使用方法

    在训练时，通过设置--load_sharded_model true --save_sharded_model true 开启reshard功能。
    load_sharded_model 表示加载的checkpoint 是否是可以自适应reshard的checkpoint，save_sharded_model 表示是否保存成可以自适应reshard的checkpoint。

    如果要支持 pp reshard, 还需要在脚本里，额外注册两个函数：
        regitser_extract_layer_name_func(extract_layer_name_func)
        register_index_layer_func(index_layer_func)
    其中，extract_layer_name_func 是一个函数，用于从参数名提取出层名；index_layer_func 是一个函数，用于从层名提取出层的 id (标识第几层)。

    该功能的具体使用方法可以参考 PR 中的demo。重点关注 llm/llama/run_pretrain.py 和 llm/llama/run_sharding_v2.sh
    这两个脚本。

### CODE REVIEW

File: paddlenlp/trainer/training_args.py [hunk 1]
```diff
@@ -710,6 +710,16 @@ class TrainingArguments:
         metadata={"help": "Whether to unify hybrid parallel checkpoint."},
     )
 
+    ignore_load_lr_and_optim: Optional[bool] = field(
```
Comment (GuoxiaWang) [inline paddlenlp/trainer/training_args.py:730]: 这个地方是不是默认为 False 更合适？

**Review by GuoxiaWang — APPROVED** [review]
LGTM `ignore_load_lr_and_optim`

File: llm/llama/run_sharding_v2.sh [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+# limitations under the License.
+
+set -x
+#unset CUDA_VISIBLE_DEVICES
+export CUDA_VISIBLE_DEVICES="0,1,2,3,4,5,6,7"
+export FLAGS_selected_gpus="0,1,2,3,4,5,6,7"
+
+rm -rf log
+rm -rf output
+#export LD_LIBRARY_PATH=/usr/local/cuda/compat:$LD_LIBRARY_PATH
```
Comment (gongweibao) [inline llm/llama/run_sharding_v2.sh]: 这个要不要根据runtime版本和驱动版本判断一下？例如
```
cuda_version=`nvidia-smi \|grep "CUDA Version" \|awk '{print $9}' \|awk -F'.' '{print $1}'`
if [ ${cuda_version} != "12" ];then
    export LD_LIBRARY_PATH=/usr/local/cuda/compat:$LD_LIBRARY_PATH
fi
```

File: llm/llama/run_sharding_v2.sh [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+    --pipeline_parallel_degree 2 \
+    --bf16  \
+    --fp16_opt_level "O2"  \
+    --amp_master_grad \
+    --tensor_parallel_degree 1 \
+    --load_sharded_model true \
+    --save_sharded_model true \
+    --ignore_data_skip 0 \
+    --force_reshard_pp true \
+    #--same_data True \
```
Comment (gongweibao) [inline llm/llama/run_sharding_v2.sh]: 这些注释掉的内容还需要吗？

File: paddlenlp/trainer/utils/sharding_io.py [hunk 4]
```diff
@@ -410,8 +471,10 @@ def _load_distributed_strategy(self, dir):
         assert "sharding_degree" in parallel_config
         return parallel_config
 
-    def _load_sharding_meta(self, dir):
-        suffix = f"tp{self.args.tensor_parallel_rank:0>2d}_pp{self.args.pipeline_parallel_rank:0>2d}"
+    def _load_sharding_meta(self, dir, pp_rank=None):
+        if pp_rank is None:
+            pp_rank = self.args.pipeline_parallel_rank
+        suffix = f"tp{self.args.tensor_parallel_rank:0>2d}_pp{pp_rank:0>2d}"
```
Comment (gongweibao) [inline paddlenlp/trainer/utils/sharding_io.py:484]: 这个地方为何要把pp_rank限定到2位？

File: paddlenlp/trainer/utils/sharding_io.py [hunk 5]
```diff
@@ -123,26 +123,46 @@ def __init__(self, args, model, optimizer=None, hcg=None):
     def set_optimizer(self, optimizer):
         self.optimizer = optimizer
 
-    def load_state_dict_from_checkpoint_with_reshard(self, resume_from_checkpoint, base_weight_name):
+    def load_state_dict_from_checkpoint_with_reshard(self, checkpoint, base_weight_name, model_wrapped):
         """load state_dict from_checkpoint with reshard, Only load model state dict."""
-        parallel_config = self._load_distributed_strategy(resume_from_checkpoint)
-        pp_degree = parallel_config["pp_degree"]
+        parallel_config = self._load_distributed_strategy(checkpoint)
```
Comment (gongweibao) [inline paddlenlp/trainer/utils/sharding_io.py:133]: 可以加一下参数的comment？

File: paddlenlp/trainer/utils/sharding_io.py (hunk 4 above)
Comment (liuzhenhai93) [reply paddlenlp/trainer/utils/sharding_io.py:484]: 哦是这样的，  目前是按 tp x pp 保存 meta 信息的

File: paddlenlp/trainer/utils/sharding_io.py [hunk 6]
```diff
@@ -123,26 +123,46 @@ def __init__(self, args, model, optimizer=None, hcg=None):
...
         pp_degree = parallel_config["pp_degree"]
         mp_degree = parallel_config["mp_degree"]
         sharding_degree = parallel_config["sharding_degree"]
-        self.args.pipeline_parallel_degree == pp_degree
-        self.args.tensor_parallel_degree == mp_degree
+        assert self.args.tensor_parallel_degree == mp_degree
```
Comment (gongweibao) [inline paddlenlp/trainer/utils/sharding_io.py]: Add explanation of this assert?
```
assert self.args.tensor_parallel_degree == mp_degree, "xxxx"
```

**Review by gongweibao — CHANGES_REQUESTED** [review]
Add some more detailed description of these
```
pipeline reshard
load param only for sharding
```

File: paddlenlp/trainer/utils/reshard/pp_reshard.py [hunk 7]
```diff
@@ -0,0 +1,332 @@
...
+
+
+def get_index_layer_func():
+    global _GLOBAL_INDEX_LAYER_FUNC
+    assert _GLOBAL_INDEX_LAYER_FUNC is not None, "index layer func is not registered yet"
+    return _GLOBAL_INDEX_LAYER_FUNC
+
+
+class LayerNameScope:
+    registered_layers = []
```
Comment (gongweibao) [inline paddlenlp/trainer/utils/reshard/pp_reshard.py:54]: Add some comments of this class ?

**Review by gongweibao — CHANGES_REQUESTED** [review]

**Review by ForFishes — APPROVED** [review]
LGTM

File: llm/llama/run_sharding_v2.sh (hunk 2 above)
Comment (liuzhenhai93) [reply llm/llama/run_sharding_v2.sh]: Done

File: llm/llama/run_sharding_v2.sh (hunk 3 above)
Comment (liuzhenhai93) [reply llm/llama/run_sharding_v2.sh]: Done

File: paddlenlp/trainer/utils/sharding_io.py (hunk 5 above)
Comment (liuzhenhai93) [reply paddlenlp/trainer/utils/sharding_io.py:133]: Done

File: paddlenlp/trainer/utils/sharding_io.py (hunk 6 above)
Comment (liuzhenhai93) [reply paddlenlp/trainer/utils/sharding_io.py]: Done

File: paddlenlp/trainer/utils/reshard/pp_reshard.py (hunk 7 above)
Comment (liuzhenhai93) [reply paddlenlp/trainer/utils/reshard/pp_reshard.py:54]: Done

**Review by gongweibao — APPROVED** [review]
LGTM
